### PR TITLE
Send archive options via pipe in chrootarchive

### LINKED ISSUE #10426: Load fails w/ FATA[0000] Error: Untar fork/exec /usr/bin/docker: argument list too long
```
$ cat /tmp/image.tar | docker load
FATA[0000] Error: Untar fork/exec /usr/bin/docker: argument list too long 
```

You can get this error if you have a several hundred images on your host.  I had ~450 images (as reported by `docker images | wc -l`) and when loading a saved image it would fail consistently w/ this error.  

This looks like it might be related to https://github.com/docker/docker/blob/092d52281d20e3456677ac11d781abf1cd0bdd8f/graph/load.go#L41 which constructs a list of all image IDs on the host which is then serialized to JSON and passed as an argument to a subsequent exec call.

After deleting my images down to ~250, `docker load` started working again.

```
$ docker version
Client version: 1.4.1
Client API version: 1.16
Go version (client): go1.3.3
Git commit (client): 5bc2ff8
OS/Arch (client): linux/amd64
Server version: 1.4.1
Server API version: 1.16
Go version (server): go1.3.3
Git commit (server): 5bc2ff8
```
```
$ docker info
Containers: 194
Images: 1526
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Dirs: 1921
Execution Driver: native-0.2
Kernel Version: 3.13.0-24-generic
Operating System: Ubuntu 14.04 LTS
CPUs: 8
Total Memory: 15.36 GiB
Name: laptop
ID: PIYC:2PIA:5YSP:XY2N:YE2Z:KO5T:YCEJ:VWQJ:MUUO:KC54:GKES:CKFT
Debug mode (server): true
Debug mode (client): false
Fds: 10
Goroutines: 10
EventsListeners: 0
Init Path: /usr/bin/docker
Docker Root Dir: /var/lib/docker
Registry: [https://index.docker.io/v1/]
```

```
$ uname -a
Linux laptop 3.13.0-24-generic #47-Ubuntu SMP Fri May 2 23:30:00 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
```

### PR DESCRIPTION
**Pull request by estesp**
Really fix #10426 this time! :smile: 

After finding our initial thinking on env. space versus arg list space
was wrong, we need to solve this by using a pipe between the caller and
child to marshall the (potentially very large) options array to the
archiver.

Docker-DCO-1.1-Signed-off-by: Phil Estes estesp@linux.vnet.ibm.com (github: estesp)

### CODE REVIEW

**Comment by LK4D4** [comment]
@crosbymichael @jfrazelle Should we add this to 1.6?

**Comment by jfrazelle** [comment]
It was a redherring!!! um ya we can

**Comment by unclejack** [comment]
Perhaps this should be in 1.6.1?

**Comment by estesp** [comment]
@LK4D4 modified to use decoders instead of marshall/unmarshall; PTAL

**Comment by jfrazelle** [comment]
yes that sounds a little less scary :)

**Comment by unclejack** [comment]
@estesp Can you add a test which breaks the old code and works on the new one, please? It's just that this would help make the case for getting this into 1.6.0 from my point of view.

**Comment by estesp** [comment]
@unclejack test is added; @LK4D4 I used the `cmd.Wait()`; wdyt?

**Comment by LK4D4** [comment]
@estesp Take a look at `CombinedOutput` implementation.

**Comment by estesp** [comment]
ooh..that's nice @LK4D4; give me another bit and I'll clean up :)

**Comment by estesp** [comment]
@LK4D4 not sure if you saw the update last evening, PTAL

File: pkg/chrootarchive/archive.go [hunk 1]
```diff
@@ -62,28 +64,39 @@ func Untar(tarArchive io.Reader, dest string, options *archive.TarOptions) error
...
+		return fmt.Errorf("Untar error on re-exec cmd: %v", err)
+	}
+	//write the options to the pipe for the untar exec to read
+	if err := json.NewEncoder(w).Encode(options); err != nil {
+		return fmt.Errorf("Untar json encode to pipe failed: %v", err)
+	}
+	w.Close()
+
+	if err := cmd.Wait(); err != nil {
+		return fmt.Errorf("Untar re-exec error: %v: output: %s", err, output.String())
```
Comment (LK4D4) [inline pkg/chrootarchive/archive.go]: You don't need t call `String` - `%s` will do this for you.

**Comment by estesp** [comment]
@LK4D4 updated, thx!

**Comment by LK4D4** [comment]
@estesp Perfect!
LGTM

**Comment by crosbymichael** [comment]
LGTM

**Comment by ehazlett** [comment]
LGTM

**Pull request merged**